src/backend/access/gin/ginfast.c

```diff
@@ -285,6 +285,15 @@ ginHeapTupleFastInsert(GinState *ginstate, GinTupleCollector *collector)
 		memset(&sublist, 0, sizeof(GinMetaPageData));
 		makeSublist(index, collector->tuples, collector->ntuples, &sublist);
 
+		if (metadata->head != InvalidBlockNumber)
+		{
+			/*
+			 * ZENITH: Get buffer before XLogBeginInsert() to avoid recursive call
+			 * of XLogBeginInsert() by zenith_wallog_page->log_newpage_copy.
+			 */
```

Comment:
Let's expand this comment a bit:
```suggestion
			 * ZENITH: Get buffer before XLogBeginInsert() to avoid recursive call
			 * of XLogBeginInsert(). Reading a new buffer might evict a dirty page from
			 * the buffer cache, and if that page happens to be an FSM or VM page, zenith_write()
			 * will try to WAL-log an image of the page.
			 */
```